### ISSUE TITLE
Custom metrics

### ISSUE THREAD

**Issue by zacfeng** [inspirehep/magpie#118]
Hi, thanks to contribute on this great tool. 
Now I have a task to develop a custom metric for muti-labeling classification, top-k categorical accuracy is too rough to evaluate since it only picks up top-K(k=5) but not weighting, for instance, prediction will be 100% accuracy even result is {a: 0.01, b: 0.01, c: 0.01} and actual is [a, b]. 

Then, custom metrics in Keras should be put in compilation step(https://keras.io/metrics/#custom_metrics), how can we do it with compile step in magpie? or any suggestion for this purpose? thanks!

**Comment by zacfeng** [comment]
Sorry my mistake, compile function for each model in`magpie/magpie/nn/models.py` do the thing. could we provide parameter for custom metric? thanks

**Comment by jstypka** [comment]
@zacfeng sounds like a reasonable thing to add. Let me work on a PR!

**Comment by jstypka** [comment]
@zacfeng I created a branch with changes you suggest. Try using it and if you're happy, we can merge it into the `master`. The branch is called `add-custom-metrics` and you can install this version of Magpie with:
```
pip install git+https://github.com/inspirehep/magpie.git@add-custom-metrics
```

Let me know how it works for you!

**Comment by zacfeng** [comment]
@jstypka you are so efficient!!!!! I found writing a metric function is not so easy.....I've tried to write a metric function by converting `y_pred` and `y_true` to numpy ndarray as follows, but it doesn't work because the metric function seems to be run as a part of tensor graph, it should be written by tensor ops and tensor operations might not suitable for this purpose(weight-based evaluation)..............

``` Python
def custom_metric(y_true, y_pred, threshold=0.5):
    normalized = np.where(K.get_value(y_pred) >= threshold, 1, 0)
    diff = np.subtract(recalculated_y_pred, K.get_value(y_true))
    rst = []
    for i, pred in enumerate(recalculated_y_pred):
        count_pred_fail = np.count_nonzero(diff[i] == 1)
        count_pred_yes = np.count_nonzero(pred == 1)
        match_rate = 0 if count_pred_yes == 0 else 1-count_pred_fail/count_pred_yes
        rst.append(match_rate)
    return K.variable(s.mean(rst))
```

The most contribution of magpie is its ease-to-use for multi-labeling,  do you have any better suggestion if someone wants to create metric without tensor? (custom callbacks may be an alternative solution for custom metric, but it's pity EarlyStopping callback would not monitor on custom metric).....

I think`add-custom-metrics` is valuable, I will keep working on writing a tensor-based metric funciton. Very appreciated for your instantly support!

**Comment by zacfeng** [comment]
By the way, I found there are lots of evaluation methods for multi-label problem in `sklearn`( [Model evaluation in sklearn](http://scikit-learn.org/stable/modules/model_evaluation.html)), maybe we can get some ideas from it.

**Comment by zacfeng** [comment]
Another idea is that could I use `ReLU` as activate function for my purpose? what I want just make output layer be `f(x) = 0 for x<0, 1 for x>=0`, but it seems not work if I change the output layer to relu.........the training result is very bad....... 

`outputs = Dense(output_length, activation='relu')(dropout)`

**Comment by jstypka** [comment]
@zacfeng the last layer is the softmax layer, so ReLU activation doesn't really fit there. You can change the activation functions in the previous layers if you wish though.

And regarding the metrics - you can reuse all the [metrics provided by Keras](https://keras.io/metrics/) and if that doesn't work, you can implement your own using Keras' backend e.g. TensorFlow. That might be more difficult, but enables you to write pretty much arbitrary logic in an efficient way.

**Comment by zacfeng** [comment]
I will keep trying and let you know if any, thank you again for your effort on this tool, it really save my time!

**Comment by zacfeng** [comment]
Sorry it took me so long to response. The code you provide in `add-custom-metrics` works fine, I add a custom metric function as follows.

```Python
def custom_metric(y_true, y_pred, threshold=0.5):
    a = K.greater_equal(y_pred, threshold)
    b = K.equal(y_true, K.cast(a, K.floatx()))
    c = K.mean(b, axis=-1)
    return c
```
So that I can specify the function outside magpie,

```Python
  hist = magpie.train(DATA_DIR,
                      labels,
                      callbacks=[earlystopping, checkpoint, csvlogger],
                      nn_model='rnn',
                      test_ratio=0.3,
                      epochs=15,
                      metrics=[custom_metric])
```

And the history will contains `custom_metric` and `val_custom_metric`, everything works find except of my metric function.....the output value is not expected for me, that is my problem I should keep going.

By the way, if I want to know more about building NN struct for text mining like this, do you have any recommendation bibble or resource to study? I've tried to replace the some Keras parameters to see it's outcome, but it usually poor, how can you achieve this fantasy nn structure design?

**Comment by jstypka** [comment]
hey, glad to hear everything worked for you! The model architecture for Magpie was inspired by a [paper by Yoon Kim](http://www.aclweb.org/anthology/D14-1181), which is worth reading. Another paper that contains tons of useful stuff on NLP might be [this one from Collobert](http://www.jmlr.org/papers/volume12/collobert11a/collobert11a.pdf), however it's pretty long and comprehensive.

If you're looking for something more lightweight and don't want to wrestle with academic writing (I don't blame you), I'd recommend [the course on Deep Learning for NLP](http://web.stanford.edu/class/cs224n/) by Richard Socher. Really great introduction and quite a lot of high-quality material.

Good luck!

**Comment by zacfeng** [comment]
Very appreciated for your help, I will read these materials!
